packages/structures/src/Structure.ts

```diff
@@ -0,0 +1,133 @@
+	}
+
+	/**
+	 * The raw data from the API for this struture
+	 *
+	 * @internal
+	 */
+	protected [kData]: Readonly<ReplaceOmittedWithUnknown<Omitted, DataType>>;
+
+	/**
+	 * Creates a new structure to represent API data
+	 *
+	 * @param data - the data from the API that this structure will represent
+	 * @param extraOptions - any additional options needed to appropriately construct a structure from the data
+	 * @remarks To be made public in subclasses
+	 * @internal
+	 */
+	protected constructor(data: Readonly<Partial<DataType>>, ..._rest: unknown[]) {
+		this[kData] = Object.assign(this._getDataTemplate(), data);
+		this[kMixinConstruct]?.(data);
+	}
+
+	/**
+	 * Patches the raw data of this object in place
+	 *
+	 * @param data - the updated data from the API to patch with
+	 * @param extraOptions - any additional options needed to appropriately patch the data
+	 * @remarks To be made public in subclasses
+	 * @returns this
+	 * @internal
+	 */
+	protected _patch(data: Readonly<Partial<DataType>>): this {
+		this[kData] = Object.assign(this._getDataTemplate(), this[kData], data);
+		this._optimizeData(data);
+		return this;
+	}
+
+	/**
+	 * Creates a clone of this structure
+	 *
+	 * @returns a clone of this
+	 * @internal
+	 */
+	protected _clone(): typeof this {
+		// @ts-expect-error constructor is of abstract class is unknown
+		return new this.constructor(this.toJSON());
+	}
+
+	/**
+	 * Function called to ensure stored raw data is in optimized formats, used in tandem with a data template
+	 *
+	 * @example created_timestamp is an ISO string, this can be stored in optimized form as a number
+	 * @param _data - the raw data received from the API to optimize
+	 * @remarks Implementation to be done in subclasses and mixins where needed.
+	 * For typescript users, mixins must use the closest ancestors access modifier.
+	 * @remarks Automatically called in {@link Structure#_patch} but must be called manually in the constructor
+	 * of any class implementing this method.
+	 * @remarks Additionally, when implementing, ensure to call `super._optimizeData` if any class in the super chain aside
+	 * from Structure contains an implementation.
+	 * Note: mixins do not need to call super ever as the process of mixing walks the prototype chain.
+	 * @virtual
+	 * @internal
+	 */
+	protected _optimizeData(_data: Partial<DataType>) {}
+
+	/**
+	 * Transforms this object to its JSON format with raw API data (or close to it),
+	 * automatically called by `JSON.stringify()` when this structure is stringified
+	 *
+	 * @remarks
+	 * The type of this data is determined by omissions at runtime and is only guaranteed for default omissions
+	 * @privateRemarks
+	 * When omitting properties at the library level, this must be overriden to re-add those properties
+	 */
+	public toJSON(): DataType {
+		// This will be DataType provided nothing is omitted, when omits occur, subclass needs to overwrite this.
+		return { ...this[kData] } as DataType;
```

Comment:
So, we have no guarantee this is actually a proper clone, right? i.e. `kData` can have objects as properties

Doesn't this fundamentally break, well, everything? `_clone()` will end up yielding references for some of those props and mutations to a totally new `_clone()`d instance will carry over to the current class as well.

We probably want as we do in builders:
```suggestion
		return structuredClone(this[kData]) as DataType;
```

Reply:
Ideally substructures (which objects as properties technically are) would be cloned in their `_toJSON()` applied after this in the Mixin. When that happens the `structuredClone` of that object would be completely replaced/overwritten again and we have unneeded overhead. So the real question is if we want to rely on implementation always overriding in Mixin or extended class if there are substructures or if we want to go the safe (but less performant) route of always deep cloning (structuredClone).

Reply:
If that's the design pattern we're opting for and we can accomplish this without `structuredClone`, then I think with some decent unit test coverage we can afford going the the override route

Reply:
worthy of note that it is entirely possible to have references when you are *not* constructing sub-structures, when you are, that data is theoretically not even present in kData anymore (to avoid double-cache)

Reply:
So we should `structuredClone()` after all?

Reply:
I'd say it makes sense to but only in cases where we need to? The problem is detecting that, it's still faster than just blindly strucutredCloning to check through values, but I don't love it